src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -366,13 +376,58 @@ void Process ()
 						if (!Annotations.IsMarked (type))
 							continue;
 						_context.MarkingHelpers.MarkExportedType (exported, assembly.MainModule, new DependencyInfo (DependencyKind.ExportedType, type));
+
+						markedExport = true;
 					}
+
+					if (markedExport)
+						ProcessDescriptorXml (assembly);
 				}
+
+				MarkFullyPreservedAssemblies ();
 			}
 
 			ProcessPendingTypeChecks ();
 		}
 
+		static bool IsFullyPreservedAction (AssemblyAction action) => action == AssemblyAction.Copy || action == AssemblyAction.Save;
+
+		void MarkFullyPreservedAssemblies ()
+		{
+			// Fully mark any assemblies with copy/save action.
+
+			// Unresolved references could get the copy/save action if this is the default action.
+			bool scanReferences = IsFullyPreservedAction (_context.CoreAction) || IsFullyPreservedAction (_context.UserAction);
+
+			if (!scanReferences) {
+				// Unresolved references could get the copy/save action if it was set explicitly
+				// for some referenced assembly that has not been resolved yet
+				foreach (DictionaryEntry e in _context.Actions) {
+					var assemblyName = (string) e.Key;
+					var action = (AssemblyAction) e.Value;
+					if (!IsFullyPreservedAction (action))
+						continue;
+
+					var assembly = _context.GetLoadedAssembly (assemblyName);
+					if (assembly == null) {
+						scanReferences = true;
+						break;
+					}
+
+					// The action should not change from the explicit command-line action
+					Debug.Assert (_context.Annotations.GetAction (assembly) == action);
+				}
+			}
+
+			var assembliesToCheck = scanReferences ? _context.ReferencedAssemblies ().ToArray () : _context.GetAssemblies ();
```

Comment:
Comment: Beware this works on loaded assemblies, not marked assemblies, so should not be tied to marking. We could do it when an assembly is loaded (since a load event is the only way to change the value of scanReferences), but we do it here to avoid doing it so frequently. We could further optimize to avoid doing this at all if the last mark iteration didn't load anything new.